cclib/method/ddec.py

```diff
@@ -271,27 +272,94 @@ def calculate(self, indices=None, fupdate=0.05):
             self.radial_grid_r.append(r)
 
         # Carry out step 3 of DDEC6 algorithm [Determine conditioned charge density and tau]
-        self.logger.info("Conditioning charge densities.")
+        self.logger.info("Conditioning charge densities. (Step 3/7)")
         self.condition_densities()
 
+        # STEP 4-7
+        # Optimize weights on each grid point
+        self._kappa = [0.0 for x in self.data.atomnos]
+        update_kappa = False
+        self.N_A = []
+        self.u_A = []
+        steps = 4
+        cycle = 0
+        while steps < 8:
+            self.logger.info("Optimizing grid weights. (Step {}/7)".format(steps))
+            self.N_A.append(self._calculate_w_and_u())
+
+            # Update kappa as described in Figure S4.2 of doi: 10.1039/c6ra04656h
+            if numpy.any([x if x < -1e-5 else 0 for x in self.N_A[-1]]) and steps != 4:
+                update_kappa = True
+            if (
+                steps > 5
+                and update_kappa
+                and numpy.diff(self.N_A)[-1] < 1e-5
+                and numpy.diff(self.N_A)[-2] < 1e-5
+            ):
+                update_kappa = False
+                self._kappa = [0.0 for x in self.data.atomnos]
+
+            if not update_kappa:
+                # If steps == 7, exit loop
+                if steps == 7:
+                    break
+                # Increment steps
+                steps = steps + 1
+                # Calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+                self.reshape_G()
+                self.calculate_H()
+
+            else:
+                # `steps` not incremented in this case
+                # First update kappa based on equation 93
+                kappa_new = self._kappa - self.N_A[cycle] / self.u_A[cycle]
+                self._kappa = [x if x > 0 else 0.0 for x in kappa_new]
+                
+            
+            # Update weights (w_A) using equation 96 in doi: 10.1039/c6ra04656h
+            for atomi in range(len(self._cond_density)):
+                self._cond_density[atomi] = math.exp(self._kappa[atomi]) * self._h[atomi]
+
+            # Update total weights on Cartesian grid using equation 65 in doi: 10.1039/c6ra04656h
+            # Generator object to iterate over the grid
+            xshape, yshape, zshape = self.chgdensity.data.shape
+            atoms = self.data.natom
+
+            self.rho_cond.data = numpy.zeros_like(self.rho_cond.data, dtype=float)
+            self._rho_cond_cartesian = numpy.zeros((atoms, xshape, yshape, zshape), dtype=float)
+
+            for atomi in range(self.data.natom):
+                grid = ((x, y, z) for x in range(xshape) for y in range(yshape) for z in range(zshape))
```

Comment:
You only use this for indexing, and I believe it's the same as iterating over `for xindex, yindex, zindex in numpy.ndindex(xshape, yshape, zshape)`.

Reply:
I've changed this. Thanks!